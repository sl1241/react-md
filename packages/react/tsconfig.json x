{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "declarationDir": "./dist/types",
    "composite": true,
    "skipLibCheck": true,
    "skipDefaultLibCheck": true,
    "jsx": "preserve",
    "jsxImportSource": "voby",
    "types": [
      "voby/dist/jsx/jsx"
    ],
  },
  "exclude": [
    "vite.config.ts",
"vite.config.umd.ts",
    "types",
  ],
}
