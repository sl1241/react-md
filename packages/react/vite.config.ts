import { defineConfig } from 'vite'

const config = defineConfig({
    build: {
        minify: false,
        lib: {
            entry: ["./src/index.ts"],
            name: "react",
            formats: ['cjs', 'es', 'umd'],
            fileName: (format: string, entryName: string) => `${entryName}.${format}.js`
        },
        // rollupOptions: {
        //   output: {
        //     preserveModules: true,
        //   }
        // },
    },


    esbuild: {
        jsx: 'automatic',
    },

    css: {
        modules: {
            scopeBehaviour: 'global',
            localsConvention: 'camelCase',
        }
    },
})



export default config
