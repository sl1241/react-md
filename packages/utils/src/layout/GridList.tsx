// import type { HTMLAttributes, Child } from 'voby'
// import { Children, forwardRef } from 'voby'

import { Children } from "@react-md/react"
import { GridListCell } from "./GridListCell"
import type { GridListSize, UseGridListOptions } from "./useGridList"
import {
    DEFAULT_GRID_LIST_MAX_CELL_SIZE,
    DEFAULT_GRID_LIST_PADDING,
    GridListSizeProvider,
    useGridList,
} from "./useGridList"

/**
 * The children render function that will be provided the current grid list size
 * object and should return renderable elements.
 *
 * Note: The first time this is called, the `columns` and `cellWidth` will be
 * the `defaultSize`.  Once the `GridList` has been fully mounted in the DOM, it
 * will begin the sizing calculations and update with the "real" values. This
 * doesn't cause any problems if you are only rendering client side, but it
 * might mess up server-side rendering, so it is recommended to update the
 * `defaultSize` when server-side rendering if this can be "known" service-side
 * in your app.
 */
export type RenderGridListChildren = (size: GridListSize) => Child

export interface GridListProps
    extends Omit<HTMLAttributes<HTMLDivElement>, "children">,
    UseGridListOptions {
    /**
     * Boolean if the resize observer should stop tracking width changes within
     * the `GridList`. This should normally stay as `false` since tracking width
     * changes will allow for dynamic content being added to the list to not mess
     * up the grid calculation when the user is on an OS that shows scrollbars.
     */
    disableHeightObserver?: FunctionMaybe<Nullable<boolean>>

    /**
     * Boolean if the resize observer should stop tracking width changes within
     * the `GridList`. This should normally stay as `false` since tracking width
     * changes will allow for dynamic content being added to the list to not mess
     * up the grid calculation when the user is on an OS that shows scrollbars.
     */
    disableWidthObserver?: FunctionMaybe<Nullable<boolean>>

    /**
     * The children to display within the grid list. This can either be a callback
     * function that will provide the current calculated width for each cell that
     * should return renderable elements or any renderable elements that are sized
     * with the `--rmd-cell-width` css variable.
     */
    children: Child | RenderGridListChildren //Child

    /**
     * Boolean if the current cell sizing should automatically be cloned into each
     * child. This will only work if the `children` is renderable element or a
     * list of renderable elements that accept the `style` and `className` props.
     */
    clone?: FunctionMaybe<Nullable<boolean>>

    /**
     * Boolean if each child within the `GridList` should be wrapped with the
     * `GridListCell` component.  This will only work if the `children` is not a
     * `function`.
     */
    wrapOnly?: FunctionMaybe<Nullable<boolean>>
}

const isRenderFunction = (
    children: GridListProps["children"]
): children is RenderGridListChildren => typeof children === "function"

/**
 * The `GridList` component is a different way to render a list of data where
 * the number of columns is dynamic and based on the max-width for each cell.
 * Instead of setting a percentage width to each cell based on the number of
 * columns, this will dynamically add columns to fill up the remaining space and
 * have each cell grow up to a set max-width. A really good use-case for this is
 * displaying a list of images or thumbnails and allowing the user to see a full
 * screen preview once selected/clicked.
 */
export const GridList = (
    {
        style,
        className,
        children,
        clone = false,
        wrapOnly = false,
        cellMargin,
        defaultSize,
        maxCellSize = DEFAULT_GRID_LIST_MAX_CELL_SIZE,
        containerPadding = DEFAULT_GRID_LIST_PADDING,
        disableHeightObserver = false,
        disableWidthObserver = false,
        ref,
        ...props
    }: GridListProps
) => {
    const [gridListProps, gridSize] = useGridList({
        //@ts-ignore
        ref, style,
        className,
        cellMargin,
        defaultSize,
        maxCellSize,
        containerPadding,
        disableHeight: disableHeightObserver,
        disableWidth: disableWidthObserver,
    })

    let content//: Child = null
    if (isRenderFunction(children)) {
        content = children(gridSize)
    } else if (clone || wrapOnly) {
        content = Children.map(
            children,
            (child) => child && <GridListCell clone={clone}>{child}</GridListCell>
        )
    } else {
        content = children
    }

    return (
        <GridListSizeProvider value={gridSize}>
            {/** @ts-ignore */}
            <div {...props} {...gridListProps}>
                {content}
            </div>
        </GridListSizeProvider>
    )
}
