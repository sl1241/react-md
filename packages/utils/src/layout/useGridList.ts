// import type { RefCallback } from 'voby'

import voby, { createContext, $, useContext, ObservableMaybe, $$ } from 'voby'
import '@react-md/react'


import { useResizeObserver } from "../sizing/useResizeObserver"
import type { PropsWithRef } from "../types"
import { useEnsuredRef } from "../useEnsuredRef"
import { scrollbarSize } from "./scrollbarSize"

/**
 * This is the css variable that is used store the current size of each cell.
 */
export const CELL_SIZE_VAR = "--rmd-cell-size"

/**
 * This is the css variable that is used store the current margin of each cell.
 */
export const CELL_MARGIN_VAR = "--rmd-cell-margin"

/**
 * @remarks \@since 2.3.0
 */
export const DEFAULT_GRID_LIST_MAX_CELL_SIZE = 150

/**
 * @remarks \@since 2.3.0
 */
export const DEFAULT_GRID_LIST_PADDING = 16

export interface GridListSize {
    /**
     * The current number of columns in the `GridList`.
     */
    columns: FunctionMaybe<number>

    /**
     * The current width of each cell within the grid.
     */
    cellWidth: FunctionMaybe<number>
}

const context = createContext<GridListSize>({
    columns: -1,
    cellWidth: -1,
})

/**
 * @remarks \@since 2.3.0
 */
export const { Provider: GridListSizeProvider } = context

if (process.env.NODE_ENV !== "production") {
    //@ts-ignore
    context.displayName = "GridListSizeProvider"
}

/**
 * Gets the current size of each cell within the `GridList` component. If this
 * is used without a parent `GridList` component, `-1` is returned instead.
 *
 * @remarks \@since 2.3.0
 */
export function useGridListSize(): GridListSize {
    return useContext(context)
}

/**
 * @remarks \@since 2.3.0
 */
export interface UseGridListOptions {
    /**
     * An optional style object to merge with the grid custom css properties
     * object.
     */
    style?: FunctionMaybe<Nullable<string | StyleProperties>>, //StyleProperties,

    /**
     * An optional className to merge with the grid list class name
     */
    className?: Class

    /**
     * Boolean if the recalculation of grid sizing should not happen for height
     * changes.
     */
    disableHeight?: FunctionMaybe<Nullable<boolean>>

    /**
     * Boolean if the recalculation of grid sizing should not happen for width
     * changes.
     */
    disableWidth?: FunctionMaybe<Nullable<boolean>>

    /**
     * An optional margin to apply to each cell as the `CELL_MARGIN_VAR` css
     * variable only when it is defined. This has to be a number string with a
     * `px`, `em`, `rem` or `%` suffix or else the grid will break.
     */
    cellMargin?: FunctionMaybe<Nullable<string>>

    /**
     * The max size that each cell can be.
     */
    maxCellSize?: FunctionMaybe<Nullable<number>>

    /**
     * Since the `GridList` requires being fully rendered in the DOM to be able to
     * correctly calculate the number of `columns` and `cellWidth`, this _might_
     * cause problems when server-side rendering when using the children renderer
     * to create a grid list dynamically based on the number of columns. If the
     * number of columns and default `cellWidth` can be guessed server-side, you
     * should provide this prop. Otherwise it will be:
     * `{ cellSize; maxCellSize, columns: -1 }`
     */
    defaultSize?: GridListSize | (() => GridListSize)

    /**
     * This is _normally_ the amount of padding on the grid list item itself to
     * subtract from the `offsetWidth` since `padding`, `border`, and vertical
     * scrollbars will be included.  If you add a border or change the padding or
     * add borders to this component, you'll need to update the `containerPadding`
     * to be the new number.
     */
    containerPadding?: FunctionMaybe<Nullable<number>>
}

export interface ProvidedGridListProps<E extends HTMLElement> {
    ref: ObservableMaybe<E | null>//RefCallback<E | null>
    style: FunctionMaybe<CSSProperties>
    className: FunctionMaybe<string>
}

type CSSProperties = voby.CSSProperties & {
    [CELL_SIZE_VAR]: string;
    [CELL_MARGIN_VAR]?: FunctionMaybe<Nullable<string>>
}

export type UseGridListReturnValue<E extends HTMLElement> = readonly [
    ProvidedGridListProps<E>,
    GridListSize
]

/**
 * The `useGridList` hook allows you to get all the grid and sizing
 * functionality of the `GridList` component without needing to wrap your
 * children in a `<div>` element.
 *
 * Example:
 *
 * ```tsx
 * const [gridListProps] = useGridList({
 *   cellMargin: 16,
 *   maxCellSize: 300,
 *   containerPadding: 4,
 * });
 *
 * return <div {...gridListProps}>{children}</div>;
 * ```
 *
 * Note: You must manually provide the `gridSize` to the `GridListSizeProvider`
 * component that was added in 2.3.0 if you want to use the `useGridSize` hook.
 *
 * Example:
 *
 * ```tsx
 * const [gridListProps, gridSize] = useGridList()
 *
 * return (
 *   <GridListSizeProvider value={gridSize}>
 *     <MyComponent {...gridListProps} />
 *   </GridListSizeProvider>
 * );
 * ```
 *
 * @remarks \@since 2.3.0
 */
export function useGridList<E extends HTMLElement>({
    ref: propRef,
    style,
    className,
    cellMargin,
    defaultSize,
    maxCellSize = DEFAULT_GRID_LIST_MAX_CELL_SIZE,
    disableHeight = false,
    disableWidth = false,
    containerPadding = DEFAULT_GRID_LIST_PADDING,
}: PropsWithRef<UseGridListOptions, E> = {}): UseGridListReturnValue<E> {
    const [ref, mergedRef] = useEnsuredRef(propRef)
    const gridSize = $((typeof defaultSize === 'function' ? defaultSize() : defaultSize) || { columns: -1, cellWidth: maxCellSize })
    const recalculate = $(() => {
        const target = ref()
        if (!target) {
            return
        }

        // need to use rect instead of offsetWidth since we need decimal precision
        // for the width since offsetWidth is basically Math.ceil(width). the
        // calculations for max columns will be off on high-pixel-density monitors
        // or some zoom levels.
        let { width } = target.getBoundingClientRect()
        width -= $$(containerPadding)

        // just need to see if there is a scrollbar visible and subtract that width.
        // don't need decimal precision here since both values will be rounded
        if (target.offsetHeight < target.scrollHeight) {
            width -= scrollbarSize("width")
        }

        const columns = Math.ceil(width / $$(maxCellSize))
        gridSize({ cellWidth: width / columns, columns })
    })
    const [, refHandler] = useResizeObserver(recalculate, {
        //@ts-ignore
        ref: mergedRef,
        disableHeight,
        disableWidth,
    })

    const mergedStyle: CSSProperties = {
        //@ts-ignore
        ...style,
        [CELL_SIZE_VAR]: `${gridSize().cellWidth}px`,
    }
    if (cellMargin) {
        mergedStyle[CELL_MARGIN_VAR] = cellMargin
    }

    return [
        {
            //@ts-ignore
            ref: refHandler,
            style: mergedStyle,
            //@ts-ignore
            className: cn("rmd-grid-list", className),
        },
        gridSize(),
    ]
}
