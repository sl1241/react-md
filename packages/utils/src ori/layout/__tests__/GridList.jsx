var __assign =
  (this && this.__assign) ||
  function () {
    __assign =
      Object.assign ||
      function (t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
          s = arguments[i];
          for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
      };
    return __assign.apply(this, arguments);
  };
import { act, render } from '@testing-library/react';
import { ResizeObserver } from '@juggle/resize-observer';
import { GridList } from '../GridList';
import { useGridListSize } from '../useGridList';
jest.mock('@juggle/resize-observer');
var ResizeObserverMock = jest.mocked(ResizeObserver);
var DEFAULT_DOM_RECT = {
  x: 100,
  y: 100,
  bottom: 2000,
  top: 100,
  left: 100,
  right: 2000,
  height: 100,
  width: 1000,
  toJSON: function () {
    return '';
  },
};
var MockedObserver = /** @class */ (function () {
  function MockedObserver(callback) {
    this._elements = [];
    this._callback = callback;
  }
  MockedObserver.prototype.observe = function (target) {
    this._elements.push(target);
  };
  MockedObserver.prototype.unobserve = function (target) {
    this._elements = this._elements.filter(function (el) {
      return el !== target;
    });
  };
  MockedObserver.prototype.disconnect = function () {
    this._elements = [];
  };
  MockedObserver.prototype.trigger = function (rect) {
    var _this = this;
    if (rect === void 0) {
      rect = {};
    }
    var contentRect = __assign(__assign({}, rect), DEFAULT_DOM_RECT);
    act(function () {
      _this._callback(
        _this._elements.map(function (target) {
          return {
            target: target,
            contentRect: contentRect,
            borderBoxSize: [],
            contentBoxSize: [],
            devicePixelContentBoxSize: [],
          };
        }),
        _this
      );
    });
  };
  return MockedObserver;
})();
var observer;
var getBoundingClientRect;
beforeAll(function () {
  ResizeObserverMock.mockImplementation(function (callback) {
    observer = new MockedObserver(callback);
    return observer;
  });
  getBoundingClientRect = jest.spyOn(
    HTMLElement.prototype,
    'getBoundingClientRect'
  );
  // this mock doesn't really matter other than the `width` value. just have to
  // provide all the rest for Typescript
  getBoundingClientRect.mockImplementation(function () {
    return __assign(__assign({}, DEFAULT_DOM_RECT), {
      toJSON: function () {
        return '';
      },
    });
  });
});
beforeEach(function () {
  observer === null || observer === void 0 ? void 0 : observer.disconnect();
});
afterAll(function () {
  ResizeObserverMock.mockRestore();
  if (getBoundingClientRect) {
    getBoundingClientRect.mockRestore();
  }
});
function trigger(width) {
  if (width === void 0) {
    width = DEFAULT_DOM_RECT.width;
  }
  act(function () {
    if (!observer) {
      throw new Error();
    }
    observer.trigger({ width: width });
  });
}
describe('GridList', function () {
  // NOTE: jsdom currently does not support rendering custom css properties (css variables)
  // so we can't actually test that part.
  it('should provide the number of columns and cellWidth to a children render function with the default of 8 padding and 150 maxCellSize', function () {
    var children = jest.fn(function (_a) {
      var columns = _a.columns,
        cellWidth = _a.cellWidth;
      return (
        <span>
          {'Columns: '.concat(columns, ', cellWidth: ').concat(cellWidth)}
        </span>
      );
    });
    var rerender = render(<GridList>{children}</GridList>).rerender;
    var containerWidth = 1000 - 16;
    expect(children).toHaveBeenCalledTimes(1);
    expect(children).toHaveBeenCalledWith({ cellWidth: 150, columns: -1 });
    trigger(containerWidth);
    expect(children).toHaveBeenCalledTimes(2);
    expect(children).toHaveBeenCalledWith({
      columns: 7,
      cellWidth: containerWidth / Math.ceil(containerWidth / 150),
    });
    children.mockClear();
    rerender(<GridList maxCellSize={400}>{children}</GridList>);
    expect(children).toHaveBeenCalledTimes(1);
    // first render then it recalculates
    expect(children).toHaveBeenCalledWith({
      columns: 7,
      cellWidth: containerWidth / Math.ceil(containerWidth / 150),
    });
    trigger(containerWidth);
    expect(children).toHaveBeenCalledWith({
      columns: 3,
      cellWidth: containerWidth / Math.ceil(containerWidth / 400),
    });
  });
  it('should provide the number of columns and cellWidth to a children render function with a custom maxCellSize', function () {
    var containerWidth = 1000 - 16;
    var children = jest.fn(function (_a) {
      var columns = _a.columns,
        cellWidth = _a.cellWidth;
      return (
        <span>
          {'Columns: '.concat(columns, ', cellWidth: ').concat(cellWidth)}
        </span>
      );
    });
    render(<GridList maxCellSize={400}>{children}</GridList>);
    expect(children).toHaveBeenCalledTimes(1);
    expect(children).toHaveBeenCalledWith({ columns: -1, cellWidth: 400 });
    trigger(containerWidth);
    expect(children).toHaveBeenCalledTimes(2);
    expect(children).toHaveBeenCalledWith({
      columns: 3,
      cellWidth: containerWidth / Math.ceil(containerWidth / 400),
    });
  });
  it('should render correctly... (lazy test)', function () {
    var _a = render(
        <GridList>
          {function (_a) {
            var columns = _a.columns,
              cellWidth = _a.cellWidth;
            return (
              <span>
                {'Columns: '.concat(columns, ', cellWidth: ').concat(cellWidth)}
              </span>
            );
          }}
        </GridList>
      ),
      container = _a.container,
      rerender = _a.rerender;
    trigger();
    expect(container).toMatchSnapshot();
    rerender(
      <GridList maxCellSize={400}>
        {function (_a) {
          var columns = _a.columns,
            cellWidth = _a.cellWidth;
          return (
            <span>
              {'Columns: '.concat(columns, ', cellWidth: ').concat(cellWidth)}
            </span>
          );
        }}
      </GridList>
    );
    trigger();
    expect(container).toMatchSnapshot();
    rerender(
      <GridList>
        <div>This is some content!</div>
      </GridList>
    );
    trigger();
    expect(container).toMatchSnapshot();
    rerender(
      <GridList maxCellSize={400}>
        <div>This is some content!</div>
      </GridList>
    );
    trigger();
    expect(container).toMatchSnapshot();
  });
  it('should allow for the current cellWidth to be accessed with the useGridListSize hook', function () {
    var Child = function () {
      var size = useGridListSize();
      return <div data-testid="child">{JSON.stringify(size)}</div>;
    };
    var getByTestId = render(
      <GridList>
        <Child />
      </GridList>
    ).getByTestId;
    trigger();
    var child = getByTestId('child');
    expect(child).toMatchInlineSnapshot(
      '\n      <div\n        data-testid="child"\n      >\n        {"cellWidth":140.57142857142858,"columns":7}\n      </div>\n    '
    );
  });
  it('should provide -1 if the useGridListSize hook is used without a parent GridList component', function () {
    var Child = function () {
      var size = useGridListSize();
      return <div data-testid="child">{JSON.stringify(size)}</div>;
    };
    var getByTestId = render(<Child />).getByTestId;
    var child = getByTestId('child');
    expect(child).toMatchInlineSnapshot(
      '\n      <div\n        data-testid="child"\n      >\n        {"columns":-1,"cellWidth":-1}\n      </div>\n    '
    );
  });
  it('should allow for all the children to be wrapped in the GridListCell component', function () {
    function Test(_a) {
      var clone = _a.clone,
        wrapOnly = _a.wrapOnly;
      return (
        <GridList clone={clone} wrapOnly={wrapOnly}>
          <div>Child 1</div>
          <div>Child 2</div>
        </GridList>
      );
    }
    var _a = render(<Test clone />),
      container = _a.container,
      rerender = _a.rerender;
    expect(container).toMatchSnapshot();
    rerender(<Test wrapOnly />);
    expect(container).toMatchSnapshot();
  });
});
//# sourceMappingURL=GridList.jsx.map
