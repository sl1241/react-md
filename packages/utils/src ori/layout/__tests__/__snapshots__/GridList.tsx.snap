// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridList should allow for all the children to be wrapped in the GridListCell component 1`] = `
<div>
  <div
    class="rmd-grid-list"
    style="--rmd-cell-size: 150px;"
  >
    <div
      class="rmd-grid-list__cell"
    >
      Child 1
    </div>
    <div
      class="rmd-grid-list__cell"
    >
      Child 2
    </div>
  </div>
</div>
`;

exports[`GridList should allow for all the children to be wrapped in the GridListCell component 2`] = `
<div>
  <div
    class="rmd-grid-list"
    style="--rmd-cell-size: 150px;"
  >
    <div
      class="rmd-grid-list__cell"
    >
      <div>
        Child 1
      </div>
    </div>
    <div
      class="rmd-grid-list__cell"
    >
      <div>
        Child 2
      </div>
    </div>
  </div>
</div>
`;

exports[`GridList should render correctly... (lazy test) 1`] = `
<div>
  <div
    class="rmd-grid-list"
    style="--rmd-cell-size: 140.57142857142858px;"
  >
    <span>
      Columns: 7, cellWidth: 140.57142857142858
    </span>
  </div>
</div>
`;

exports[`GridList should render correctly... (lazy test) 2`] = `
<div>
  <div
    class="rmd-grid-list"
    style="--rmd-cell-size: 328px;"
  >
    <span>
      Columns: 3, cellWidth: 328
    </span>
  </div>
</div>
`;

exports[`GridList should render correctly... (lazy test) 3`] = `
<div>
  <div
    class="rmd-grid-list"
    style="--rmd-cell-size: 140.57142857142858px;"
  >
    <div>
      This is some content!
    </div>
  </div>
</div>
`;

exports[`GridList should render correctly... (lazy test) 4`] = `
<div>
  <div
    class="rmd-grid-list"
    style="--rmd-cell-size: 328px;"
  >
    <div>
      This is some content!
    </div>
  </div>
</div>
`;
